apollo-router-core/src/context.rs

```diff
@@ -44,13 +46,59 @@ impl From<Context<http_compat::Request<Request>>> for Context<Arc<http_compat::R
 }
 
 impl<T> Context<T> {
-    pub fn extensions(&self) -> impl Future<Output = RwLockReadGuard<Object>> {
+    /// Take a read lock of extensions
+    /// Be careful when using this method, you could create deadlock if you use `lock_extensions_mut` in the same scope
+    pub fn lock_extensions(&self) -> impl Future<Output = RwLockReadGuard<Object>> {
         self.extensions.read()
     }
 
-    pub fn extensions_mut(&self) -> impl Future<Output = RwLockWriteGuard<Object>> {
+    /// Take a write lock of extensions
+    /// Be careful when using this method, you could create deadlock if you use `lock_extensions` in the same scope
+    pub fn lock_extensions_mut(&self) -> impl Future<Output = RwLockWriteGuard<Object>> {
         self.extensions.write()
     }
```

Comment:
Can we consider going further and remove these two functions completely?

Reply:
I'm not sure TBH. Maybe sometimes people will need to play with IterMut or Iter or something equivalent. Will ask tomorrow during the huddle if someone have thoughts